homeassistant/components/fritzbox_callmonitor/sensor.py

```diff
@@ -20,95 +26,131 @@
-        vol.Optional(CONF_USERNAME, default=""): cv.string,
-        vol.Optional(CONF_PHONEBOOK, default=0): cv.positive_int,
-        vol.Optional(CONF_PREFIXES, default=[]): vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PHONEBOOK, default=DEFAULT_PHONEBOOK): cv.positive_int,
+        vol.Optional(CONF_PREFIXES): vol.All(cv.ensure_list, [cv.string]),
     }
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up Fritz!Box call monitor sensor platform."""
-    name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
-    # Try to resolve a hostname; if it is already an IP, it will be returned as-is
-    try:
-        host = socket.gethostbyname(host)
-    except OSError:
-        _LOGGER.error("Could not resolve hostname %s", host)
-        return
-    port = config.get(CONF_PORT)
-    username = config.get(CONF_USERNAME)
-    password = config.get(CONF_PASSWORD)
-    phonebook_id = config.get(CONF_PHONEBOOK)
-    prefixes = config.get(CONF_PREFIXES)
-
-    try:
-        phonebook = FritzBoxPhonebook(
-            host=host,
-            port=port,
-            username=username,
-            password=password,
-            phonebook_id=phonebook_id,
-            prefixes=prefixes,
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Import the platform into a config entry."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
         )
-    except:  # noqa: E722 pylint: disable=bare-except
-        phonebook = None
-        _LOGGER.warning("Phonebook with ID %s not found on Fritz!Box", phonebook_id)
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the fritzbox_callmonitor sensor from config_entry."""
+    phonebook = hass.data[DOMAIN][config_entry.entry_id][FRITZ_BOX_PHONEBOOK_OBJECT]
 
-    sensor = FritzBoxCallSensor(name=name, phonebook=phonebook)
+    phonebook_name = config_entry.title
+    phonebook_id = config_entry.data[CONF_PHONEBOOK]
+    prefixes = config_entry.options.get(CONF_PREFIXES)
+    serial_number = config_entry.data[SERIAL_NUMBER]
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
 
-    add_entities([sensor])
+    phonebook_info = await hass.async_add_executor_job(
+        phonebook.fph.phonebook_info, phonebook_id
+    )
+    actual_phonebook_name = phonebook_info[FRITZ_ATTR_NAME]
 
-    monitor = FritzBoxCallMonitor(host=host, port=port, sensor=sensor)
-    monitor.connect()
+    if phonebook_name != actual_phonebook_name:
+        name = phonebook_name
+    else:
+        name = f"{phonebook.fph.modelname} Call Monitor {phonebook_name}"
```

Comment:
Uncertain here as well regarding how far it has to go to provide backwards compatibility for import from yaml to config entry. This solution is not perfect either but I tried to make the transition as smooth as possible. Especially difficult since I would prefer to drop the ability to choose the name of a created entity via config flow.

Reply:
Just drop it from the config flow and for imports, set some value in `entry.data`. Instead of setting a name to impact the resulting entity ID, you can also just set `entity_id`. It will be used as a suggestion if it's the first time the entity is seen.

Reply:
Done in https://github.com/home-assistant/core/pull/40736/commits/7359e27e6d2e6facce99b2861d1d48dc20993078